### PR TITLE
Change TFMs for most projects based on project type and output type

### PR DESCRIPTION
**Pull request by cartermp**
* Exes go to highest app TFM by default
* Libraries go to NS 2.0 by default
* Respect `tfm` for everything if it was specified
* Some refactoring

### CODE REVIEW

**Comment by jmarolf** [comment]
>     public class MSBuildWorkspace

Can we _please_ rename this to something like `ProjectComparer` every time I see this type name I get confused 

---
Refers to: src/MSBuild.Abstractions/MSBuildWorkspace.cs:14 in 1170e81. [](commit_id = 1170e819b086bb1a4e067460c7ba210794ef5a2c, deletion_comment = False)

File: src/MSBuild.Abstractions/MSBuildWorkspace.cs [hunk 1]
```diff
@@ -157,7 +158,42 @@ private BaselineProject CreateSdkBaselineProject(string projectFilePath, IProjec
...
+            {
+                ProjectOutputType.Exe => true,
+                ProjectOutputType.Library => true,
+                ProjectOutputType.WinExe => true,
+                _ => false
+            };
+
+        private ProjectOutputType GetProjectOutputType(IProjectRootElement root)
+        {
+            var outputTypeNode = root.GetOutputTypeNode();
```
Comment (jmarolf) [inline src/MSBuild.Abstractions/MSBuildWorkspace.cs:175]: why no switch expression?

File: src/MSBuild.Conversion.Project/Converter.cs [hunk 2]
```diff
@@ -24,20 +25,18 @@ public Converter(UnconfiguredProject project, BaselineProject sdkBaselineProject
             _differs = GetDiffers();
         }
 
-        public void Convert(string outputPath, string? defaultTFM, bool keepCurrentTfm)
+        public void Convert(string outputPath, string? specifiedTFM, bool keepCurrentTfm, bool usePreviewSDK)
         {
-            ConvertProjectFile(defaultTFM, keepCurrentTfm);
-            CleanUpProjectFile(outputPath);
+            var result = ConvertProjectFile(specifiedTFM, keepCurrentTfm, usePreviewSDK);
+            if (result is { })
```
Comment (jmarolf) [inline src/MSBuild.Conversion.Project/Converter.cs:31]: I have so much the giant baby

Comment (jmarolf) [reply src/MSBuild.Conversion.Project/Converter.cs:31]: please just use `result != null` like a normal person

---
In reply to: [422375425](https://github.com/dotnet/try-convert/pull/291#discussion_r422375425) [](ancestors = 422375425)

File: src/MSBuild.Conversion.Project/Converter.cs [hunk 3]
```diff
@@ -55,6 +54,33 @@ internal IProjectRootElement ConvertProjectFile(string? defaultTFM, bool keepCur
...
+            static string GetBestTFM(BaselineProject baselineProject, bool keepCurrentTfm, string? specifiedTFM, bool usePreviewSDK)
+            {
+                if (string.IsNullOrWhiteSpace(specifiedTFM))
+                {
+                    // Let's figure this out, friends
+                    var tfmForApps = TargetFrameworkHelper.FindHighestInstalledTargetFramework(usePreviewSDK);
+
+                    specifiedTFM =
+                        keepCurrentTfm
+                        ? baselineProject.GetTfm()
```
Comment (jmarolf) [inline src/MSBuild.Conversion.Project/Converter.cs:67]: image

Comment (jmarolf) [reply src/MSBuild.Conversion.Project/Converter.cs:67]: Can we use some `if` statements? ternmary straigh into a double nested switch for one of the more complex code paths is going to be hard to debug

---
In reply to: [422376966](https://github.com/dotnet/try-convert/pull/291#discussion_r422376966) [](ancestors = 422376966)

File: src/MSBuild.Conversion.Project/Converter.cs (hunk 2 above)
Comment (cartermp) [reply src/MSBuild.Conversion.Project/Converter.cs:31]: actually I can just get rid of this nonsense, it's an artifact from something else I did but ended up not doing

File: src/MSBuild.Abstractions/MSBuildWorkspace.cs (hunk 1 above)
Comment (cartermp) [reply src/MSBuild.Abstractions/MSBuildWorkspace.cs:175]: This is a `ProjectPropertyElement?` so it's not switch-able

File: src/MSBuild.Abstractions/ProjectItemComparer.cs [hunk 4]
```diff
@@ -15,8 +15,23 @@ private ProjectItemComparer(bool compareMetadata)
             _compareMetadata = compareMetadata;
         }
 
-        public bool Equals(IProjectItem x, IProjectItem y)
+        public bool Equals(IProjectItem? x, IProjectItem? y)
```
Comment (jmarolf) [inline src/MSBuild.Abstractions/ProjectItemComparer.cs:18]: You can keep these non-nullable and add `[AllowNull]` in front of each to indicate that the intent is that you not pass null

File: src/MSBuild.Conversion.Project/ProjectRootElementExtensionsForConversion.cs [hunk 5]
```diff
@@ -11,8 +10,9 @@
 
 namespace MSBuild.Conversion.Project
 {
-    public static class ProjectRootElementExtensions
+    public static class ProjectRootElementExtensionsForConversion
     {
+
```
Comment (jmarolf) [inline src/MSBuild.Conversion.Project/ProjectRootElementExtensionsForConversion.cs:15]: just say no to blank lines

File: src/MSBuild.Abstractions/ProjectItemComparer.cs (hunk 4 above)
Comment (cartermp) [reply src/MSBuild.Abstractions/ProjectItemComparer.cs:18]: Yeah that’s preferable

File: src/MSBuild.Conversion.Project/ProjectRootElementExtensionsForConversion.cs (hunk 5 above)
Comment (cartermp) [reply src/MSBuild.Conversion.Project/ProjectRootElementExtensionsForConversion.cs:15]: image

**Pull request merged**